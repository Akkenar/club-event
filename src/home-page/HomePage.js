import React, { useEffect } from 'react';
import InformationSection from '../information-section/InformationSection';
import { setPageTitle } from '../page.lib';
import getKey from '../intl/getKey';
import { injectIntl } from 'react-intl';

const HomePage = props => {
  useEffect(() => {
    setPageTitle(getKey('home.page.title', props));
  });

  return (
    <div>
      <InformationSection />
    </div>
  );
};

export default injectIntl(HomePage);
