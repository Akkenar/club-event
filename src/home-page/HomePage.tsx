import React from 'react';
import { Fragment, useContext, useEffect } from 'react';
import { Header } from 'semantic-ui-react';
import imgTop from '../assets/Arnaud-Conne-2017_5.jpg';
import getKey from '../core/intl/getKey';
import LanguageContext from '../core/intl/LanguageContext';
import LazyImage from '../core/lazy-image/LazyImage';
import Logo from '../core/logo/Logo';
import { goToTop, setPageTitle } from '../core/page.lib';
import InformationSection from './information-section/InformationSection';
import ScrollableAnchor from '../core/scrollable-anchor/ScrollableAnchor';
import MapIframe from './MapIframe';
import ScheduleSection from './schedule-section/ScheduleSection';
import SponsorsSection from './sponsors-section/SponsorsSection';
import RegisterSection from './register-section/RegisterSection';
import PageMenu from './PageMenu';

const HomePage = () => {
  const { messages } = useContext(LanguageContext);

  useEffect(() => {
    setPageTitle(getKey('home.page.title', messages));
    goToTop();
  });

  return (
    <Fragment>
      <PageMenu />
      <Header as="h1" data-testid="main-title">
        <ScrollableAnchor text={getKey('home.page.title', messages)} />
      </Header>
      <Logo big={true} width={320} height={232} />
      <InformationSection />
      <MapIframe />
      <ScheduleSection />
      <SponsorsSection />
      <RegisterSection />
      <LazyImage
        width={320}
        height={212}
        src={imgTop}
        alt={'Photo: Arnaud Conne'}
      />
    </Fragment>
  );
};

export default HomePage;
