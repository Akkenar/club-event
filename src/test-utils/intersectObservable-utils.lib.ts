import { mockGlobalProperty } from './test-utils.lib';

const allObservers: IntersectionObserverMock[] = [];

class IntersectionObserverMock {
  private readonly cb: any;
  public constructor(cb: any) {
    this.cb = cb;

    // To allow global triggering.
    allObservers.push(this);
  }

  public trigger() {
    this.cb([{ isIntersecting: true }]);
  }

  public observe() {
    // Nothing
  }

  public unobserve() {
    // Nothing.
  }

  public disconnect() {
    // Nothing.
  }
}

class IntersectionObserverEntryMock {
  public intersectionRatio() {
    // Nothing.
  }

  public isIntersecting() {
    // Nothing.
  }
}

export function mockInterestObservable(): () => void {
  // Reset all the mocks
  allObservers.slice(0);

  mockGlobalProperty(window)('IntersectionObserver')(IntersectionObserverMock);
  mockGlobalProperty(window)('IntersectionObserverEntry')(
    IntersectionObserverEntryMock,
  );

  // Trigger all.
  return () => {
    allObservers.forEach((observer: IntersectionObserverMock) => {
      observer.trigger();
    });
  };
}
