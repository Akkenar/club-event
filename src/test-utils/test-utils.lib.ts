import { fireEvent } from '@testing-library/react';

export function mockGrecaptcha() {
  (window as any).grecaptcha = {
    getResponse: () => 'response',
    render: jest.fn(),
    reset: jest.fn(),
  };
}

export function mockResponse(response: any) {
  const mock = typeof response === 'string' ? { result: response } : response;

  // Mock the fetch data as JSDom doesn't implement fetch.
  // @ts-ignore
  window.fetch = jest.fn(() =>
    Promise.resolve({
      text: () =>
        new Promise(resolve => {
          const mockResult = JSON.stringify(mock);
          setTimeout(() => {
            resolve(mockResult);
          }, 50);
        }),
    }),
  );
}

export function setValidInputData(
  validData: any,
  { getByTestId }: { getByTestId: any },
) {
  Object.keys(validData).forEach((key: string) => {
    fireEvent.change(getByTestId(key), {
      // @ts-ignore
      target: { value: validData[key] },
    });
  });
}

export const mockGlobalProperty = (globalObject: any) => (key: string) => (
  value: any,
) => {
  const original = globalObject[key];
  Object.defineProperty(globalObject, key, { value, writable: true });
  return () => mockGlobalProperty(globalObject)(key)(original);
};

export function forceLoadImage() {
  Object.defineProperty(Image.prototype, 'onload', {
    configurable: true,
    set(fn: () => void) {
      // Call the function directly to simulate the loading
      fn();
    },
  });
}
