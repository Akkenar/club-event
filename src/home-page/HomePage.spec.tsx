import * as React from 'react';
import { BrowserRouter } from 'react-router-dom';
import { act, render, waitForElement, wait } from '@testing-library/react';
import { mockInterestObservable } from '../test-utils/intersectObservable-utils.lib';
import { forceLoadImage } from '../test-utils/test-utils.lib';
import HomePage from './HomePage';

const HomePageWithRouter = () => (
  <BrowserRouter>
    <HomePage />
  </BrowserRouter>
);

forceLoadImage();

describe('HomePage', () => {
  let showAsyncElements: () => void;
  beforeEach(() => {
    showAsyncElements = mockInterestObservable();
  });

  it('should match the snapshot', async () => {
    const wrapper = render(<HomePageWithRouter />);

    act(() => {
      showAsyncElements();
    });

    // Wait for the logo.
    await waitForElement(() => wrapper.queryByTitle('logo.alt'));

    // Wait for the iframe.
    await waitForElement(() => wrapper.getByTitle('Vallorbe'));

    expect(wrapper.baseElement).toMatchSnapshot();
  });

  it('should set the page title', () => {
    // Will be set by side effect through a hook
    render(<HomePageWithRouter />);
    expect(document.title).toEqual('home.page.title');
  });

  it('should navigate with the page menu', async () => {
    window.scrollTo = jest.fn();
    const wrapper = render(<HomePageWithRouter />);

    act(() => {
      showAsyncElements();
    });

    // One of the menu item.
    await wrapper.getByTestId('home.menu.register').click();

    await wait(() => {
      // The focus should be on the element.
      expect(document.activeElement?.innerHTML).toEqual('home.page.register');
      expect(window.scrollTo).toHaveBeenCalledTimes(1);
    });
  });
});
