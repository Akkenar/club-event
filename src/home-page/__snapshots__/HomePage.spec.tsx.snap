// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage should match the snapshot 1`] = `
<body>
  <div>
    <h1
      class="ui header"
      data-testid="main-title"
    >
      home.page.title
    </h1>
    <img
      alt="logo.alt"
      class="Logo  Logo--big"
      data-testid="logo"
      height="232"
      src="[object Object]"
      title="logo.alt"
      width="320"
    />
    <div />
    <iframe
      allowfullscreen=""
      frameborder="0"
      height="600"
      src="https://www.google.com/maps/embed?pb=!1m18!1m12!1m3!1d2735.705437069773!2d6.377488315962078!3d46.71154597913546!2m3!1f0!2f0!3f0!3m2!1i1024!2i768!4f13.1!3m3!1m2!1s0x478db3ee662c90db%3A0x172582dbb0e2b277!2sAuberge+Communale+de+Vallorbe!5e0!3m2!1sen!2sch!4v1541839403749"
      style="border: 0px; margin: 1em 0px;"
      title="Vallorbe"
      width="100%"
    />
    <h2
      class="ui header"
    >
      home.page.register
    </h2>
    <div
      class="InformationSection__register-container"
    >
      <a
        class="ui black button full-width"
        data-testid="register-button"
        href="/undefined/register"
        role="button"
      >
        register.now
      </a>
    </div>
    <div
      class="LazyImage no-print "
    >
      <picture
        style="display: block;"
      >
        <source
          srcset="test-file-stub"
          type="image/webp"
        />
        <img
          alt="Photo: Arnaud Conne"
          class="LazyImage__Image"
          data-testid="lazyloaded-image"
          height="212"
          src="test-file-stub"
          title="Photo: Arnaud Conne"
          width="320"
        />
      </picture>
      <div
        aria-hidden="true"
        class="LazyImage__Legend"
      >
        Photo: Arnaud Conne
      </div>
    </div>
  </div>
</body>
`;
